clang/lib/Parse/ParseTemplate.cpp

```diff
@@ -337,6 +343,8 @@ Parser::ParseConceptDefinition(const ParsedTemplateInfo &TemplateInfo,
   ExprResult ConstraintExprResult =
       Actions.CorrectDelayedTyposInExpr(ParseConstraintExpression());
   if (ConstraintExprResult.isInvalid()) {
+    if (AddedToScope)
```

Comment:
I really don't like the `AddedToScope` pass by reference and I am wondering why we can't do this inside of `ActOnStartConceptDefinition`

Reply:
I'm not sure I understand it, but did you mean to move the checking to the Sema despite it relying on the constraint expression having been parsed?

Reply:
Oh NVM, I see your point:
`PushOnScopeChains()` is hinged on `Previous.empty()`, so I'm afraid we need the flag, otherwise we can definitely do away with the flag

Reply:
Can we just mark the decl invalid?

Reply:
Oh yes - they're not evaluated either that way.